File: leetcode/javaStuff/sortsAndSearches/quickSort/QuickSort.java
// O(1) space
// O(nlog(n)) time

public class QuickSort {
  public static void main(String[] args) {
      int[] test = {6,2,12,3,1,36,2};

      quickSort(test);

      for(int i = 0; i < test.length; i++) {
        System.out.println(test[i]);
      }
  }

  public static void quickSort(int[] arr) {
    quickSort(arr, 0, arr.length - 1);
  }

  public static void quickSort(int[] arr, int left, int right) {
    if(left >= right) {
      return;
    }

    int pivot = arr[(left + right) / 2];
    int index = partition(arr, left, right, pivot);
    quickSort(arr, left, index - 1);
    quickSort(arr, index, right);


  }

  // Does all the swapping and then returns the point which divides
  // all the elements lower than the pivot and all the elements greater
  // than the pivot
  public static int partition(int[] arr, int left, int right, int pivot) {

    while(left <= right) {

      // find left which belongs on right side
      while(arr[left] < pivot) {
        left++;
      }

      // find right which belongs on left side
      while(arr[right] > pivot) {
        right--;
      }

      // once both of those values are found, swap them thangs
      if(left <= right) {
        int temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;
        left++;
        right--;
      }


    }

    return left;

  }
}

File: leetcode/sortsAndSearchesJS/quick_sort.js
// booty take up hella space way for practice
function quickSort(arr) {
  if(arr.length <= 1) {
    return arr;
  }

  let partition = arr.shift();
  let left = [];
  let right = [];

  arr.forEach( el => {
    if(el <= partition) {
      left.push(el);
    } else {
      right.push(el);
    }
  });

  return quickSort(left).concat([partition]).concat(quickSort(right));
}

function quickSort2(arr, left, right) {
  if(left >= right) {
    return arr;
  }

  let pivot = arr[right];
  let partitionIndex = partition(arr, left, right, pivot);

  // console.log(partitionIndex);
  // console.log(arr);

  quickSort2(arr, left, partitionIndex - 1);
  quickSort2(arr, partitionIndex + 1, right);
}

function partition(arr, left, right, pivot) {
  let partitionIndex = left;

  for(let i = left; i < right; i++) {
    if(arr[i] < pivot) {
      swap(arr, i , partitionIndex);
      partitionIndex++;
    }
  }

  swap(arr, partitionIndex, right);
  return partitionIndex;
}

function swap(arr, a, b) {
  let temp = arr[a];
  arr[a] = arr[b];
  arr[b] = temp;

}

let farts = [5,23,1,8,56,2,9,546,13,12,5,8,9];
quickSort2(farts, 0, farts.length - 1);
console.log(farts);
